File: pix/src/main/java/com/fxn/adapters/MainImageAdapter.java
package com.fxn.adapters;

import android.content.Context;
import android.net.Uri;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.RequestManager;
import com.bumptech.glide.load.DecodeFormat;
import com.bumptech.glide.load.engine.DiskCacheStrategy;
import com.bumptech.glide.request.RequestOptions;
import com.fxn.interfaces.OnSelectionListener;
import com.fxn.interfaces.SectionIndexer;
import com.fxn.modals.Img;
import com.fxn.pix.R;
import com.fxn.utility.HeaderItemDecoration;
import com.fxn.utility.Utility;

import java.io.File;
import java.util.ArrayList;

/**
 * Created by akshay on 17/03/18.
 */

public class MainImageAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> implements HeaderItemDecoration.StickyHeaderInterface, SectionIndexer {

    public static final int HEADER = 1;
    public static final int ITEM = 2;
    public static int SPAN_COUNT = 0;
    private static final int MARGIN = 4;

    private final ArrayList<Img> list;
    private OnSelectionListener onSelectionListener;
    private final FrameLayout.LayoutParams layoutParams;
    private final RequestManager glide;
    private final RequestOptions options;

    public MainImageAdapter(Context context, int spanCount) {
        this.list = new ArrayList<>();
        SPAN_COUNT = spanCount;
        int size = (Utility.WIDTH / SPAN_COUNT) - (MARGIN / 2);
        layoutParams = new FrameLayout.LayoutParams(size, size);
        layoutParams.setMargins(MARGIN, MARGIN - (MARGIN / 2), MARGIN, MARGIN - (MARGIN / 2));
        options = new RequestOptions().override(size - 50)
                .format(DecodeFormat.PREFER_RGB_565)
                .centerCrop()
                .diskCacheStrategy(DiskCacheStrategy.RESOURCE);
        glide = Glide.with(context);
    }

    public ArrayList<Img> getItemList() {
        return list;
    }

    public MainImageAdapter addImage(Img image) {
        list.add(image);
        notifyDataSetChanged();
        return this;
    }

    public void addOnSelectionListener(OnSelectionListener onSelectionListener) {
        this.onSelectionListener = onSelectionListener;
    }

    public void addImageList(ArrayList<Img> images) {
        list.addAll(images);
        notifyDataSetChanged();
    }

    @Override
    public int getItemViewType(int position) {
	    if (list.size() <= position) return 0;
        Img i = list.get(position);
        return (i.getContentUrl().equalsIgnoreCase("")) ? HEADER : ITEM;
    }

    public void clearList() {
        list.clear();
    }

    public void select(boolean selection, int pos) {
        list.get(pos).setSelected(selection);
        notifyItemChanged(pos);
    }

    @Override
    public long getItemId(int position) {
        return list.get(position).getContentUrl().hashCode();
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        if (viewType == HEADER) {
            return new HeaderHolder(LayoutInflater.from(parent.getContext())
                    .inflate(R.layout.header_row, parent, false));
        } else {
            View view = LayoutInflater.from(parent.getContext())
                    .inflate(R.layout.main_image, parent, false);
            return new Holder(view);
        }
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        Img image = list.get(position);
        if (holder instanceof Holder) {
            Holder imageHolder = (Holder) holder;
            if (image.getMedia_type() == 1) {
                glide.load(image.getContentUrl()).apply(options).into(imageHolder.preview);
                imageHolder.isVideo.setVisibility(View.GONE);
            } else if (image.getMedia_type() == 3) {
                glide.asBitmap()
                        .load(Uri.fromFile(new File(image.getUrl())))
                        .apply(options)
                        .into(imageHolder.preview);
                imageHolder.isVideo.setVisibility(View.VISIBLE);
            }
            imageHolder.selection.setVisibility(image.getSelected() ? View.VISIBLE : View.GONE);
        } else if (holder instanceof HeaderHolder) {
            HeaderHolder headerHolder = (HeaderHolder) holder;
            headerHolder.header.setText(image.getHeaderDate());
        }
    }

    @Override
    public int getItemCount() {
        return list.size();
    }

    @Override
    public int getHeaderPositionForItem(int position) {
        int itemPosition = position;
        int headerPosition = 0;
        do {
            if (this.isHeader(itemPosition)) {
                headerPosition = itemPosition;
                break;
            }
            itemPosition -= 1;
        } while (itemPosition >= 0);
        return headerPosition;
    }

    @Override
    public int getHeaderLayout(int headerPosition) {
        return R.layout.header_row;
    }

    @Override
    public void bindHeaderData(View header, int headerPosition) {
        Img image = list.get(headerPosition);
        ((TextView) header.findViewById(R.id.header)).setText(image.getHeaderDate());
    }

    @Override
    public boolean isHeader(int itemPosition) {
        return getItemViewType(itemPosition) == 1;
    }

    @Override
    public String getSectionText(int position) {
        return list.get(position).getHeaderDate();
    }

    public String getSectionMonthYearText(int position) {
	    if (list.size() <= position) return "";
        return list.get(position).getHeaderDate();
    }

    public class Holder extends RecyclerView.ViewHolder implements View.OnClickListener, View.OnLongClickListener {
        private final ImageView preview;
        private final ImageView selection;
        private final ImageView isVideo;

        Holder(View itemView) {
            super(itemView);
            preview = itemView.findViewById(R.id.preview);
            selection = itemView.findViewById(R.id.selection);
            isVideo = itemView.findViewById(R.id.isVideo);
            itemView.setOnClickListener(this);
            itemView.setOnLongClickListener(this);
            preview.setLayoutParams(layoutParams);
        }

        @Override
        public void onClick(View view) {
            int id = this.getLayoutPosition();
            onSelectionListener.onClick(list.get(id), view, id);
        }

        @Override
        public boolean onLongClick(View view) {
            int id = this.getLayoutPosition();
            onSelectionListener.onLongClick(list.get(id), view, id);
            return true;
        }
    }

    public static class HeaderHolder extends RecyclerView.ViewHolder {
        private final TextView header;

        HeaderHolder(View itemView) {
            super(itemView);
            header = itemView.findViewById(R.id.header);
        }
    }
}

File: pix/src/main/java/com/fxn/adapters/MainImageAdapter.kt
package com.fxn.adapters

import android.content.Context
import android.net.Uri
import android.view.LayoutInflater
import android.view.View
import android.view.View.OnLongClickListener
import android.view.ViewGroup
import android.widget.FrameLayout
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.bumptech.glide.RequestManager
import com.bumptech.glide.load.DecodeFormat
import com.bumptech.glide.load.engine.DiskCacheStrategy
import com.bumptech.glide.request.RequestOptions
import com.fxn.interfaces.OnSelectionListener
import com.fxn.interfaces.SectionIndexer
import com.fxn.modals.Img
import com.fxn.pix.R
import com.fxn.utility.HeaderItemDecoration.StickyHeaderInterface
import com.fxn.utility.Utility
import java.io.File
import java.util.*

/**
 * Created by akshay on 17/03/18.
 */
class MainImageAdapter(context: Context, spanCount: Int) : RecyclerView.Adapter<RecyclerView.ViewHolder>(), StickyHeaderInterface, SectionIndexer {
    val itemList: ArrayList<Img> = ArrayList()
    private var onSelectionListener: OnSelectionListener? = null
    private val layoutParams: FrameLayout.LayoutParams
    private val glide: RequestManager
    private val options: RequestOptions

    fun addImage(image: Img): MainImageAdapter {
        itemList.add(image)
        notifyDataSetChanged()
        return this
    }

    fun addOnSelectionListener(onSelectionListener: OnSelectionListener?) {
        this.onSelectionListener = onSelectionListener
    }

    fun addImageList(images: ArrayList<Img>?) {
        itemList.addAll(images!!)
        notifyDataSetChanged()
    }

    override fun getItemViewType(position: Int): Int {
        if (itemList.size <= position) {
            return 0
        }
        val i = itemList[position]
        return if (i.contentUrl.equals("", ignoreCase = true)) HEADER else ITEM
    }

    fun clearList() {
        itemList.clear()
    }

    fun select(selection: Boolean, pos: Int) {
        itemList[pos].selected = selection
        notifyItemChanged(pos)
    }

    override fun getItemId(position: Int): Long {
        return itemList[position].contentUrl.hashCode().toLong()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return if (viewType == HEADER) {
            HeaderHolder(LayoutInflater.from(parent.context).inflate(R.layout.header_row, parent, false))
        } else {
            val view = LayoutInflater.from(parent.context).inflate(R.layout.main_image, parent, false)
            Holder(view)
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val image = itemList[position]
        if (holder is Holder) {
            if (image.media_type == 1) {
                glide.load(image.contentUrl).apply(options).into(holder.preview)
                holder.isVideo.visibility = View.GONE
            } else if (image.media_type == 3) {
                glide.asBitmap()
                        .load(Uri.fromFile(File(image.url)))
                        .apply(options)
                        .into(holder.preview)
                holder.isVideo.visibility = View.VISIBLE
            }
            holder.selection.visibility = if (image.selected) View.VISIBLE else View.GONE
        } else if (holder is HeaderHolder) {
            holder.header.text = image.headerDate
        }
    }

    override fun getItemCount(): Int {
        return itemList.size
    }

    override fun getHeaderPositionForItem(itemPosition: Int): Int {
        var position = itemPosition
        var headerPosition = 0
        do {
            if (isHeader(position)) {
                headerPosition = position
                break
            }
            position -= 1
        } while (position >= 0)
        return headerPosition
    }

    override fun getHeaderLayout(headerPosition: Int): Int {
        return R.layout.header_row
    }

    override fun bindHeaderData(header: View?, headerPosition: Int) {
        val image = itemList[headerPosition]
        (header!!.findViewById<View>(R.id.header) as TextView).text = image.headerDate
    }

    override fun isHeader(itemPosition: Int): Boolean {
        return getItemViewType(itemPosition) == 1
    }

    override fun getSectionText(position: Int): String {
        return itemList[position].headerDate
    }

    fun getSectionMonthYearText(position: Int): String {
        return if (itemList.size <= position) {
            ""
        } else itemList[position].headerDate
    }

    inner class Holder internal constructor(itemView: View) : RecyclerView.ViewHolder(itemView), View.OnClickListener, OnLongClickListener {
        val preview: ImageView = itemView.findViewById(R.id.preview)
        val selection: ImageView = itemView.findViewById(R.id.selection)
        val isVideo: ImageView = itemView.findViewById(R.id.isVideo)
        override fun onClick(view: View) {
            val id = this.layoutPosition
            onSelectionListener!!.onClick(itemList[id], view, id)
        }

        override fun onLongClick(view: View): Boolean {
            val id = this.layoutPosition
            onSelectionListener!!.onLongClick(itemList[id], view, id)
            return true
        }

        init {
            itemView.setOnClickListener(this)
            itemView.setOnLongClickListener(this)
            preview.layoutParams = layoutParams
        }
    }

    inner class HeaderHolder internal constructor(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val header: TextView = itemView.findViewById(R.id.header)
    }

    companion object {
        const val HEADER = 1
        const val ITEM = 2
        var SPAN_COUNT = 0
        private const val MARGIN = 4
    }

    init {
        SPAN_COUNT = spanCount
        val size = Utility.WIDTH / SPAN_COUNT - MARGIN / 2
        layoutParams = FrameLayout.LayoutParams(size, size)
        layoutParams.setMargins(MARGIN, MARGIN - MARGIN / 2, MARGIN, MARGIN - MARGIN / 2)
        options = RequestOptions().override(size - 50)
                .format(DecodeFormat.PREFER_RGB_565)
                .centerCrop()
                .diskCacheStrategy(DiskCacheStrategy.RESOURCE)
        glide = Glide.with(context)
    }
}